libraries/chain/controller.cpp

```diff
@@ -3999,15 +3986,99 @@ struct controller_impl {
+      EOS_ASSERT( !f_ext.new_proposer_policy_diff, invalid_qc_claim,
+                  "Transition block #${b} has new_proposer_policy_diff",
+                  ("b", block_num) );
+
+      if (auto it = prev.header_exts.find(finality_extension::extension_id()); it != prev.header_exts.end()) {
+         const auto& prev_finality_ext = std::get<finality_extension>(it->second);
+         EOS_ASSERT( f_ext.qc_claim.block_num == prev_finality_ext.qc_claim.block_num,
+                     invalid_qc_claim,
+                     "Transition block #${b} QC claim block_num not equal to previous QC claim block_num",
+                     ("b", block_num) );
+      } else {
+         // Savanna Genesis Block
+         EOS_ASSERT( f_ext.qc_claim.block_num == block_num, invalid_qc_claim,
+                     "Savanna Genesis block #${b} QC claim block_num not equal to current block_num",
+                     ("b", block_num) );
+         EOS_ASSERT( f_ext.new_finalizer_policy_diff, invalid_qc_claim,
+                     "Savanna Genesis block #${b} finality block header extension misses new_finalizer_policy_diff",
+                     ("b", block_num) );
+
+         finalizer_policy no_policy;
+         // apply_diff will FC_ASSERT if new_finalizer_policy_diff is malformated
+         finalizer_policy genesis_policy = no_policy.apply_diff(*f_ext.new_finalizer_policy_diff);
+         EOS_ASSERT( genesis_policy.generation == 1, invalid_qc_claim,
+                     "Savanna Genesis block #${b} finalizer policy generation (${g}) not 1",
+                     ("b", block_num)("g", genesis_policy.generation) );
+      }
+   }
+
    // thread safe, expected to be called from thread other than the main thread
    template<typename ForkDB, typename BS>
    block_handle create_block_state_i( ForkDB& forkdb, const block_id_type& id, const signed_block_ptr& b, const BS& prev ) {
       constexpr bool savanna_mode = std::is_same_v<typename std::decay_t<BS>, block_state>;
+
+      // Verify claim made by finality_extension in block header extension and
+      // quorum_certificate_extension in block extension are valid.
+      // This is the only place the evaluation is done.
+      // Note: the purpose of running verify_qc_claim on Legacy blocks too is to
+      // safe guard Legacy blocks.
       if constexpr (savanna_mode) {
-         // Verify claim made by finality_extension in block header extension and
-         // quorum_certificate_extension in block extension are valid.
-         // This is the only place the evaluation is done.
          verify_qc_claim(id, b, prev);
+      } else {
+         EOS_ASSERT( !b->is_proper_svnn_block(), block_validate_exception,
+                     "create_block_state_i cannot be called on block #${b} which is not a Proper Savanna block while its parent is a Legacy block or a Transition block",
+                     ("b", b->block_num()) );
+
+         verify_legacy_and_transition_block_exts_common(b, prev);
```

Comment:
I suggest removing this and duplicating the checks in `verify_legacy_block_exts` and `verify_transition_block_exts`. That way you can be more specific in the error messages.